src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/RazorProjectChangePublisher.cs

```diff
@@ -0,0 +1,45 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Concurrent;
+using System.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.Razor.ProjectSystem
+{
+    internal abstract class RazorProjectChangePublisher : ProjectSnapshotChangeTrigger
+    {
+        private ProjectSnapshotManagerBase _projectSnapshotManager;
+        protected readonly ConcurrentDictionary<string, string> _publishFilePathMappings;
+
+        public RazorProjectChangePublisher()
+        {
+            _publishFilePathMappings = new ConcurrentDictionary<string, string>(FilePathComparer.Instance);
+        }
+
+        public override void Initialize(ProjectSnapshotManagerBase projectManager)
+        {
+            _projectSnapshotManager = projectManager;
+            _projectSnapshotManager.Changed += ProjectSnapshotManager_Changed;
+        }
+
+        internal abstract bool IsLSPEditorAvailable(string projectFilePath);
+
+        internal abstract void ProjectSnapshotManager_Changed(object sender, ProjectChangeEventArgs e);
+
+        protected abstract bool IsOnMainThread();
```

Comment:
Why make this abstract?

Reply:
We `Debug.Assert` that we're on the main thread in `SetPublishFilePath` and `RemovePublishFilePath`, if we pass JoinableTaskContext down to `RazorProjectChangePublisher` then we have to add `Microsoft.VisualStudio.Threading` to `Microsoft.CodeAnalysis.Razor.Workspaces`, which doesn't seem like it should contain anything VS specific. I can add a comment to this effect.